File: apps/api/schemas/destiny.py
"""
命理进阶功能 Pydantic 响应模型定义
"""

from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


# ============================================================
# 大运流年
# ============================================================

class LuckPeriod(BaseModel):
    """大运周期"""
    start_age: int = Field(..., description="起始年龄")
    end_age: int = Field(..., description="结束年龄")
    heavenly_stem: str = Field(..., description="天干")
    earthly_branch: str = Field(..., description="地支")
    ganzhi: str = Field(..., description="干支")
    element: str = Field(..., description="五行")
    luck_level: str = Field(..., description="旺衰等级：旺/相/休/囚/死")


class MajorLuckResponse(BaseModel):
    """大运周期响应"""
    luck_periods: List[LuckPeriod] = Field(..., description="大运周期列表")
    current_luck: Optional[LuckPeriod] = Field(None, description="当前大运")


class AnnualLuck(BaseModel):
    """流年运势"""
    year: int = Field(..., description="年份")
    heavenly_stem: str = Field(..., description="天干")
    earthly_branch: str = Field(..., description="地支")
    ganzhi: str = Field(..., description="干支")
    element: str = Field(..., description="五行")
    relationship: str = Field(..., description="与日主关系")
    advice: str = Field(..., description="流年建议")


class AnnualLuckResponse(BaseModel):
    """流年运势响应"""
    annual_luck: AnnualLuck = Field(..., description="流年信息")
    scores: Dict[str, int] = Field(..., description="五维度评分")
    overall_score: int = Field(..., description="综合评分")
    lucky_colors: List[str] = Field(default_factory=list, description="幸运颜色")
    lucky_materials: List[str] = Field(default_factory=list, description="幸运材质")
    lucky_directions: List[str] = Field(default_factory=list, description="幸运方位")
    lucky_elements: List[str] = Field(default_factory=list, description="幸运五行")
    outfit_advice: str = Field(..., description="穿搭建议")


# ============================================================
# 十神
# ============================================================

class TenGodInfo(BaseModel):
    """单柱十神信息"""
    stem: str = Field(..., description="天干")
    ganzhi: str = Field(..., description="干支")
    ten_god: str = Field(..., description="十神名称")


class ShenShaInfo(BaseModel):
    """命带神煞信息"""
    name: str = Field(..., description="神煞名")
    category: str = Field(..., description="分类：吉/中性/煞")
    positions: List[str] = Field(default_factory=list, description="出现柱位")
    duanyu: str = Field(..., description="传统断语")


class TenGodsResponse(BaseModel):
    """十神格局分析响应"""
    pillars: Dict[str, TenGodInfo] = Field(..., description="四柱十神")
    hidden_gods: List[Dict[str, Any]] = Field(default_factory=list, description="藏干十神")
    dominant_gods: List[str] = Field(default_factory=list, description="旺神")
    weak_gods: List[str] = Field(default_factory=list, description="衰神")
    god_distribution: Dict[str, float] = Field(default_factory=dict, description="十神分布")
    analysis: str = Field(..., description="格局分析")
    shen_sha: List[ShenShaInfo] = Field(default_factory=list, description="命带神煞")
    shen_sha_note: str = Field("", description="神煞合规角标文案")


# ============================================================
# 月度运势
# ============================================================

class MonthlyFortuneResponse(BaseModel):
    """月度运势响应"""
    year: int = Field(..., description="年份")
    month: int = Field(..., description="月份")
    month_ganzhi: str = Field(..., description="月柱干支")
    scores: Dict[str, int] = Field(..., description="五维度评分")
    overall_score: int = Field(..., description="综合评分")
    element_analysis: Dict[str, Any] = Field(..., description="五行分析")
    outfit_strategy: Dict[str, Any] = Field(..., description="穿搭策略")


class MonthlySummary(BaseModel):
    """月度运势摘要"""
    month: int = Field(..., description="月份")
    month_ganzhi: str = Field(..., description="月柱干支")
    overall_score: int = Field(..., description="综合评分")
    scores: Dict[str, int] = Field(..., description="五维度评分")
    outfit_strategy: Dict[str, Any] = Field(default_factory=dict, description="穿搭策略")


class YearlyFortuneResponse(BaseModel):
    """年度运势响应"""
    year: int = Field(..., description="年份")
    overall_score: int = Field(..., description="年度综合评分")
    monthly_summary: List[MonthlySummary] = Field(..., description="12月运势摘要")
    peak_months: List[int] = Field(default_factory=list, description="旺月")
    low_months: List[int] = Field(default_factory=list, description="衰月")
    yearly_advice: str = Field(..., description="年度建议")


# ============================================================
# 高级八字分析
# ============================================================

class AdvancedBaziResponse(BaseModel):
    """高级八字分析响应"""
    pillars: Dict[str, str] = Field(..., description="四柱")
    nayin: Dict[str, Any] = Field(..., description="纳音五行")
    hidden_stems: Dict[str, Any] = Field(..., description="地支藏干")
    chong: Dict[str, Any] = Field(..., description="冲")
    xing: Dict[str, Any] = Field(..., description="刑")
    hai: Dict[str, Any] = Field(..., description="害")
    he: Dict[str, Any] = Field(..., description="合")
    analysis: str = Field(..., description="综合分析")

File: apps/web/store/destiny.ts
/**
 * 命理分析状态管理
 * 十神格局、大运流年、高级八字分析
 */
import { create } from 'zustand'
import { getTenGods, getAnnualLuck, getMajorLuck, getAdvancedBazi } from '@/lib/api'

// ========== 类型定义 ==========

export interface ShenShaInfo {
  name: string
  category: '吉' | '中性' | '煞' | string
  positions: string[]
  duanyu: string
}

export interface TenGodsData {
  pillars: Record<string, { stem: string; ganzhi: string; ten_god: string }>
  hidden_gods: Array<{ pillar: string; hidden_stem: string; ten_god: string }>
  dominant_gods: string[]
  weak_gods: string[]
  god_distribution: Record<string, number>
  analysis: string
  shen_sha?: ShenShaInfo[]
  shen_sha_note?: string
}

export interface LuckPeriod {
  start_age: number
  end_age: number
  heavenly_stem: string
  earthly_branch: string
  ganzhi: string
  element: string
  luck_level: string
}

export interface AnnualLuckData {
  annual_luck: {
    year: number
    heavenly_stem: string
    earthly_branch: string
    ganzhi: string
    element: string
    relationship: string
    advice: string
  }
  scores: Record<string, number>
  overall_score: number
  lucky_colors: string[]
  lucky_materials: string[]
  lucky_directions: string[]
  lucky_elements: string[]
  outfit_advice: string
}

export interface NayinItem {
  ganzhi: string
  nayin_name: string
  nayin_element: string
  nayin_description: string
}

export interface HiddenStemItem {
  stem: string
  element: string
  is_main: boolean
}

export interface AdvancedBaziData {
  pillars: Record<string, string>
  nayin: Record<string, NayinItem>
  hidden_stems: Record<string, HiddenStemItem[]>
  chong: { has_chong: boolean; pairs: Array<{ branch_a: string; branch_b: string; description: string }>; count: number }
  xing: { has_xing: boolean; groups: Array<{ type: string; branches: string[]; description: string }>; count: number }
  hai: { has_hai: boolean; pairs: Array<{ branch_a: string; branch_b: string; description: string }>; count: number }
  he: { has_he: boolean; sanhe: Array<{ type: string; branches: string[]; element: string; description: string }>; liuhe: Array<{ branch_a: string; branch_b: string; element: string; description: string }>; count: number }
  analysis: string
}

// ========== Store ==========

interface DestinyState {
  // 十神格局
  tenGods: TenGodsData | null
  isTenGodsLoading: boolean
  tenGodsError: string | null

  // 大运周期
  majorLuck: { luck_periods: LuckPeriod[]; current_luck: LuckPeriod | null } | null
  isMajorLuckLoading: boolean
  majorLuckError: string | null

  // 流年运势
  annualLuck: AnnualLuckData | null
  isAnnualLuckLoading: boolean
  annualLuckError: string | null

  // 高级八字分析
  advancedBazi: AdvancedBaziData | null
  isAdvancedBaziLoading: boolean
  advancedBaziError: string | null

  // Actions
  fetchTenGods: () => Promise<void>
  fetchMajorLuck: () => Promise<void>
  fetchAnnualLuck: (year?: number) => Promise<void>
  fetchAdvancedBazi: () => Promise<void>
  fetchAll: () => Promise<void>
  clearErrors: () => void
}

export const useDestinyStore = create<DestinyState>()((set, get) => ({
  // 十神
  tenGods: null,
  isTenGodsLoading: false,
  tenGodsError: null,

  // 大运
  majorLuck: null,
  isMajorLuckLoading: false,
  majorLuckError: null,

  // 流年
  annualLuck: null,
  isAnnualLuckLoading: false,
  annualLuckError: null,

  // 高级八字
  advancedBazi: null,
  isAdvancedBaziLoading: false,
  advancedBaziError: null,

  fetchTenGods: async () => {
    set({ isTenGodsLoading: true, tenGodsError: null })
    try {
      set({ tenGods: await getTenGods(), isTenGodsLoading: false })
    } catch (e) {
      set({ tenGodsError: e instanceof Error ? e.message : '获取十神格局失败', isTenGodsLoading: false })
    }
  },

  fetchMajorLuck: async () => {
    set({ isMajorLuckLoading: true, majorLuckError: null })
    try {
      set({ majorLuck: await getMajorLuck(), isMajorLuckLoading: false })
    } catch (e) {
      set({ majorLuckError: e instanceof Error ? e.message : '获取大运周期失败', isMajorLuckLoading: false })
    }
  },

  fetchAnnualLuck: async (year?: number) => {
    set({ isAnnualLuckLoading: true, annualLuckError: null })
    try {
      set({ annualLuck: await getAnnualLuck(year || new Date().getFullYear()), isAnnualLuckLoading: false })
    } catch (e) {
      set({ annualLuckError: e instanceof Error ? e.message : '获取流年运势失败', isAnnualLuckLoading: false })
    }
  },

  fetchAdvancedBazi: async () => {
    set({ isAdvancedBaziLoading: true, advancedBaziError: null })
    try {
      set({ advancedBazi: await getAdvancedBazi(), isAdvancedBaziLoading: false })
    } catch (e) {
      set({ advancedBaziError: e instanceof Error ? e.message : '获取高级八字分析失败', isAdvancedBaziLoading: false })
    }
  },

  fetchAll: async () => {
    const { fetchTenGods, fetchMajorLuck, fetchAnnualLuck, fetchAdvancedBazi } = get()
    fetchTenGods()
    fetchMajorLuck()
    fetchAnnualLuck()
    fetchAdvancedBazi()
  },

  clearErrors: () => set({
    tenGodsError: null,
    majorLuckError: null,
    annualLuckError: null,
    advancedBaziError: null,
  }),
}))
